File: sdk-ts/src/base-client.ts
/**
 * Base client implementation with shared logic for HTTP requests and error handling
 */

import { VERSION } from './version.js';
import {
  AuthenticationError,
  PermissionError,
  NotFoundError,
  ValidationError,
  RateLimitError,
  APIError,
} from './exceptions.js';

/**
 * Configuration options for the base client
 */
export interface BaseClientConfig {
  /** Mnemosyne API key (required, or set MNEMOSYNE_API_KEY env var) */
  apiKey?: string;
  /** Base URL for API (default: http://localhost:8000/api/v1). Include /api/v1 suffix when using custom URLs */
  baseUrl?: string;
  /** Request timeout in milliseconds (default: 60000 = 60 seconds) */
  timeout?: number;
  /** Maximum number of retries for failed requests (default: 3) */
  maxRetries?: number;
}

/**
 * Request options for HTTP requests
 */
interface RequestOptions {
  json?: unknown;
  params?: Record<string, string | number | boolean>;
  headers?: Record<string, string>;
}

/**
 * Base client with shared logic for HTTP requests and error handling.
 *
 * Provides:
 * - HTTP client management with native fetch
 * - Authentication headers
 * - Error handling and exception mapping
 * - Retry logic with exponential backoff
 */
export class BaseClient {
  /** @internal API key for authentication */
  public readonly apiKey: string;
  /** @internal Base URL for API requests */
  public readonly baseUrl: string;
  /** @internal Request timeout in milliseconds */
  public readonly timeout: number;
  /** @internal Maximum retry attempts */
  public readonly maxRetries: number;

  constructor(config: BaseClientConfig = {}) {
    // Get API key from config or environment
    this.apiKey = config.apiKey || process.env.MNEMOSYNE_API_KEY || '';

    if (!this.apiKey) {
      throw new Error(
        'API key is required. Provide it in config or set MNEMOSYNE_API_KEY environment variable.'
      );
    }

    // Get base URL from config or environment
    // Default includes /api/v1 prefix for API routes
    const rawBaseUrl =
      config.baseUrl || process.env.MNEMOSYNE_BASE_URL || 'http://localhost:8000/api/v1';

    // Remove trailing slash (matching Python SDK behavior)
    // This allows string concatenation with leading-slash paths
    this.baseUrl = rawBaseUrl.endsWith('/') ? rawBaseUrl.slice(0, -1) : rawBaseUrl;

    this.timeout = config.timeout || 60000; // 60 seconds default
    this.maxRetries = config.maxRetries || 3;
  }

  /**
   * Get authentication headers for requests
   *
   * @param options - Header options
   * @param options.skipAuth - Skip Authorization header (default: false)
   * @param options.skipContentType - Skip Content-Type header for multipart uploads (default: false)
   * @returns Headers object
   */
  protected getHeaders(options: {
    skipAuth?: boolean;
    skipContentType?: boolean;
  } = {}): Record<string, string> {
    const headers: Record<string, string> = {};

    // Add Authorization unless skipped
    if (!options.skipAuth) {
      headers['Authorization'] = `Bearer ${this.apiKey}`;
    }

    // Add Content-Type unless skipped (for multipart uploads)
    if (!options.skipContentType) {
      headers['Content-Type'] = 'application/json';
    }

    // Always add User-Agent
    headers['User-Agent'] = `mnemosyne-typescript/${VERSION}`;

    return headers;
  }

  /**
   * Handle error responses asynchronously with proper error message extraction
   * @internal
   */
  public async handleErrorAsync(response: Response): Promise<void> {
    if (response.ok) {
      return;
    }

    const statusCode = response.status;
    let message = `HTTP ${statusCode} error`;

    // Try to extract error message from response
    try {
      const errorData = (await response.json()) as { detail?: string };
      message = errorData.detail || response.statusText || message;
    } catch {
      message = response.statusText || message;
    }

    // Map status codes to exception types
    switch (statusCode) {
      case 401:
        throw new AuthenticationError(message, statusCode);
      case 403:
        throw new PermissionError(message, statusCode);
      case 404:
        throw new NotFoundError(message, statusCode);
      case 422:
        throw new ValidationError(message, statusCode);
      case 429:
        throw new RateLimitError(message, statusCode);
      default:
        throw new APIError(message, statusCode);
    }
  }

  /**
   * Determine if a request should be retried
   */
  protected shouldRetry(response: Response | null, error: Error | null): boolean {
    // Retry on network errors
    if (error) {
      return true;
    }

    // Retry on 5xx server errors and 429 rate limits
    if (response && (response.status === 429 || response.status >= 500)) {
      return true;
    }

    return false;
  }

  /**
   * Calculate exponential backoff delay in milliseconds
   * @returns Delay in milliseconds (2^attempt seconds, max 16 seconds)
   */
  protected calculateBackoff(attempt: number): number {
    return Math.min(Math.pow(2, attempt) * 1000, 16000); // Max 16 seconds (16000ms)
  }

  /**
   * Make an HTTP request with retry logic
   */
  async request<T>(
    method: string,
    path: string,
    options: RequestOptions = {},
    skipAuth = false
  ): Promise<T> {
    // Construct URL using string concatenation (like Python SDK)
    // This allows leading slashes in paths while supporting /api/v1 in baseUrl
    const urlString = path.startsWith('http') ? path : this.baseUrl + path;
    const url = new URL(urlString);

    // Add query parameters
    if (options.params) {
      for (const [key, value] of Object.entries(options.params)) {
        url.searchParams.append(key, String(value));
      }
    }

    // Prepare headers
    const headers = {
      ...this.getHeaders({ skipAuth }),
      ...options.headers,
    };

    let lastError: Error | null = null;

    for (let attempt = 0; attempt < this.maxRetries; attempt++) {
      try {
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), this.timeout);

        const response = await fetch(url.toString(), {
          method,
          headers,
          body: options.json ? JSON.stringify(options.json) : undefined,
          signal: controller.signal,
        });

        clearTimeout(timeoutId);

        // Check for errors and decide whether to retry
        if (!response.ok) {
          // Determine if we should retry this error
          if (this.shouldRetry(response, null) && attempt < this.maxRetries - 1) {
            // Will retry - wait with backoff
            const delay = this.calculateBackoff(attempt);
            await new Promise((resolve) => setTimeout(resolve, delay));
            continue;
          }

          // Not retryable or out of retries - throw error with proper message
          await this.handleErrorAsync(response);
        }

        // Success - parse and return response
        return (await response.json()) as T;
      } catch (error) {
        lastError = error as Error;

        // If it's an HTTP error, don't retry
        if (
          error instanceof AuthenticationError ||
          error instanceof PermissionError ||
          error instanceof NotFoundError ||
          error instanceof ValidationError
        ) {
          throw error;
        }

        // Retry on network errors
        if (attempt < this.maxRetries - 1) {
          const delay = this.calculateBackoff(attempt);
          await new Promise((resolve) => setTimeout(resolve, delay));
          continue;
        }
      }
    }

    // All retries exhausted
    throw new APIError(
      `Request failed after ${this.maxRetries} retries: ${lastError?.message || 'Unknown error'}`
    );
  }

  /**
   * Make an HTTP request with multipart/form-data body and retry logic
   *
   * This method is specifically for file uploads and handles FormData properly
   * while maintaining retry logic and error handling.
   */
  async requestMultipart<T>(method: string, path: string, formData: FormData): Promise<T> {
    // Construct URL
    const urlString = path.startsWith('http') ? path : this.baseUrl + path;
    const url = new URL(urlString);

    // Get headers WITHOUT Content-Type (browser sets it automatically for FormData)
    const headers = this.getHeaders({ skipContentType: true });

    let lastError: Error | null = null;

    for (let attempt = 0; attempt < this.maxRetries; attempt++) {
      try {
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), this.timeout);

        const response = await fetch(url.toString(), {
          method,
          headers,
          body: formData,
          signal: controller.signal,
        });

        clearTimeout(timeoutId);

        // Check for errors and decide whether to retry
        if (!response.ok) {
          if (this.shouldRetry(response, null) && attempt < this.maxRetries - 1) {
            const delay = this.calculateBackoff(attempt);
            await new Promise((resolve) => setTimeout(resolve, delay));
            continue;
          }

          // Not retryable or out of retries
          await this.handleErrorAsync(response);
        }

        // Success
        return (await response.json()) as T;
      } catch (error) {
        lastError = error as Error;

        // If it's an HTTP error, don't retry
        if (
          error instanceof AuthenticationError ||
          error instanceof PermissionError ||
          error instanceof NotFoundError ||
          error instanceof ValidationError
        ) {
          throw error;
        }

        // Retry on network errors
        if (attempt < this.maxRetries - 1) {
          const delay = this.calculateBackoff(attempt);
          await new Promise((resolve) => setTimeout(resolve, delay));
          continue;
        }
      }
    }

    // All retries exhausted
    throw new APIError(
      `Request failed after ${this.maxRetries} retries: ${lastError?.message || 'Unknown error'}`
    );
  }

  /**
   * Make a streaming request (returns Response for SSE processing)
   */
  async requestStream(
    method: string,
    path: string,
    options: RequestOptions = {},
    skipAuth = false
  ): Promise<Response> {
    // Construct URL using string concatenation (like Python SDK)
    const urlString = path.startsWith('http') ? path : this.baseUrl + path;
    const url = new URL(urlString);

    // Add query parameters
    if (options.params) {
      for (const [key, value] of Object.entries(options.params)) {
        url.searchParams.append(key, String(value));
      }
    }

    // Prepare headers for SSE
    const headers = {
      ...this.getHeaders({ skipAuth }),
      Accept: 'text/event-stream',
      ...options.headers,
    };

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeout);

    const response = await fetch(url.toString(), {
      method,
      headers,
      body: options.json ? JSON.stringify(options.json) : undefined,
      signal: controller.signal,
    });

    clearTimeout(timeoutId);

    // Handle errors
    await this.handleErrorAsync(response);

    return response;
  }
}

File: sdk/mnemosyne/_base_client.py
"""Base client implementation with shared logic for sync and async clients"""

import time
from typing import Optional, Any, Dict
import httpx
from .version import __version__
from .exceptions import (
    AuthenticationError,
    PermissionError,
    NotFoundError,
    ValidationError,
    RateLimitError,
    APIError,
)


class BaseClient:
    """
    Base client with shared logic for HTTP requests and error handling.

    This class provides:
    - HTTP client management
    - Authentication headers
    - Error handling and exception mapping
    - Retry logic with exponential backoff
    """

    def __init__(
        self,
        api_key: str,
        base_url: str = "http://localhost:8000/api/v1",
        timeout: float = 60.0,
        max_retries: int = 3,
    ):
        """
        Initialize base client.

        Args:
            api_key: Mnemosyne API key (required)
            base_url: Base URL for API (default: http://localhost:8000/api/v1)
            timeout: Request timeout in seconds (default: 60.0)
            max_retries: Maximum number of retries for failed requests (default: 3)

        Raises:
            ValueError: If api_key is empty
        """
        if not api_key:
            raise ValueError("api_key is required")

        self.api_key = api_key
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        self.max_retries = max_retries

        # Will be set by subclasses
        self._http_client: Optional[httpx.Client] = None

    def _get_headers(self, include_content_type: bool = True) -> Dict[str, str]:
        """
        Get authentication and default headers for requests

        Args:
            include_content_type: Whether to include Content-Type: application/json
                                 (default: True, set to False for multipart uploads)

        Returns:
            Dict of headers including Authorization and User-Agent
        """
        headers = {
            "Authorization": f"Bearer {self.api_key}",
            "User-Agent": f"mnemosyne-python/{__version__}",
        }

        if include_content_type:
            headers["Content-Type"] = "application/json"

        return headers

    def _handle_error(self, response: httpx.Response) -> None:
        """
        Handle error responses and raise appropriate exceptions.

        Args:
            response: HTTP response object

        Raises:
            AuthenticationError: For 401 responses
            PermissionError: For 403 responses
            NotFoundError: For 404 responses
            ValidationError: For 422 responses
            RateLimitError: For 429 responses
            APIError: For other error responses
        """
        if response.is_success:
            return

        status_code = response.status_code
        try:
            error_data = response.json()
            message = error_data.get("detail", response.text)
        except Exception:
            message = response.text or f"HTTP {status_code} error"

        # Map status codes to exception types
        if status_code == 401:
            raise AuthenticationError(message, status_code)
        elif status_code == 403:
            raise PermissionError(message, status_code)
        elif status_code == 404:
            raise NotFoundError(message, status_code)
        elif status_code == 422:
            raise ValidationError(message, status_code)
        elif status_code == 429:
            raise RateLimitError(message, status_code)
        else:
            raise APIError(message, status_code)

    def _should_retry(self, response: Optional[httpx.Response], exception: Optional[Exception]) -> bool:
        """
        Determine if a request should be retried.

        Args:
            response: HTTP response (if available)
            exception: Exception raised (if any)

        Returns:
            bool: True if request should be retried
        """
        # Retry on network errors
        if exception and isinstance(exception, (httpx.RequestError, httpx.TimeoutException)):
            return True

        # Retry on 5xx server errors and 429 rate limits
        if response and response.status_code in (429, 500, 502, 503, 504):
            return True

        return False

    def _calculate_backoff(self, attempt: int) -> float:
        """
        Calculate exponential backoff delay.

        Args:
            attempt: Current retry attempt (0-indexed)

        Returns:
            float: Delay in seconds
        """
        return min(2 ** attempt, 16)  # Max 16 seconds

    def _prepare_request_url(self, path: str) -> str:
        """
        Prepare full URL from path

        Args:
            path: Request path (e.g., "/collections") or full URL

        Returns:
            str: Full URL for request
        """
        return path if path.startswith("http") else f"{self.base_url}{path}"

    def _should_retry_attempt(self, attempt: int, response: Optional[httpx.Response], exception: Optional[Exception]) -> bool:
        """
        Determine if we should retry and how long to wait

        Args:
            attempt: Current retry attempt (0-indexed)
            response: HTTP response (if available)
            exception: Exception raised (if any)

        Returns:
            bool: True if should retry (and not on last attempt)
        """
        if attempt >= self.max_retries - 1:
            return False

        return self._should_retry(response, exception)

    def close(self) -> None:
        """Close the HTTP client"""
        if self._http_client:
            self._http_client.close()

    def __enter__(self):
        """Context manager entry"""
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        """Context manager exit"""
        self.close()
        return False
